yaml/service.go

```diff
@@ -58,21 +60,78 @@ func (s *ServiceSlice) ToPipeline() *pipeline.ContainerSlice {
 // UnmarshalYAML implements the Unmarshaler interface for the ServiceSlice type.
 //
 //nolint:dupl // accepting duplicative code that exists in step.go as well
-func (s *ServiceSlice) UnmarshalYAML(unmarshal func(interface{}) error) error {
+func (s *ServiceSlice) UnmarshalYAML(v *yaml.Node) error {
+	// service slice should be sequence
+	if v.Kind != yaml.SequenceNode {
+		return fmt.Errorf("invalid yaml: expected sequence node for service slice")
+	}
+
 	// service slice we try unmarshalling to
 	serviceSlice := new([]*Service)
 
-	// attempt to unmarshal as a service slice type
-	err := unmarshal(serviceSlice)
-	if err != nil {
-		return err
-	}
-
 	// iterate through each element in the service slice
-	for _, service := range *serviceSlice {
-		// handle nil service to avoid panic
-		if service == nil {
-			return fmt.Errorf("invalid service with nil content found")
+	for _, st := range v.Content {
+		// make local var
+		tmpService := *st
+
+		// services are mapping nodes
+		if tmpService.Kind != yaml.MappingNode {
+			return fmt.Errorf("invalid yaml: expected map node for service")
+		}
+
+		// initialize anchor node -- will be nil if `<<` never used
+		var anchorKey *yaml.Node
+
+		// initialize anchor sets
+		anchorList := new([]*yaml.Node)  // collect multiple anchor references
+		newContent := new([]*yaml.Node)  // new service content
+		anchorSequence := new(yaml.Node) // final type that is appended to service contents
+
+		// iterate through map contents (key, value)
+		for i := 0; i < len(tmpService.Content); i += 2 {
+			key := tmpService.Content[i]
+			value := tmpService.Content[i+1]
+
+			// check if key is an anchor reference
+			if strings.EqualFold(key.Value, "<<") {
```

Comment:
EqualFold() seems unnecessary here (and in the other location).

Reply:
if the equalfold is necessary you could use a switch case comparison for some extra speed 
```
switch key.Value {
case "<<":
```

Reply:
i think it's a simple `key.Value == "<<"`, no? there are no casing concerns.

Reply:
right, switch is faster tho (jump tables) but its negligible (with over 1 billion iterations it measures like 3% faster)